src/main/java/com/epam/aidial/core/storage/BlobWriteStream.java

```diff
@@ -112,17 +115,21 @@ public void end(Handler<AsyncResult<Void>> handler) {
                     throw new RuntimeException(exception);
                 }
 
-                byte[] lastChunk = chunkBuffer.slice(0, position).getBytes();
+                ByteBuf lastChunk = chunkBuffer.slice(0, position).getByteBuf();
                 if (mpu == null) {
                     log.info("Resource is too small for multipart upload, sending as a regular blob");
-                    metadata = resourceService.putFile(resource, lastChunk, etag, contentType);
+                    try (InputStream chunkStream = new ByteBufInputStream(lastChunk)) {
+                        metadata = resourceService.putFile(resource, chunkStream.readAllBytes(), etag, contentType);
+                    }
                 } else {
                     if (position != 0) {
-                        MultipartPart part = storage.storeMultipartPart(mpu, ++chunkNumber, lastChunk);
-                        parts.add(part);
+                        try (InputStream chunkStream = new ByteBufInputStream(lastChunk.copy())) {
```

Comment:
Why do you copy it?

Reply:
You can use duplicate to create a shallow copy or just reset reader index before calling etag calculation.

Reply:
Reading changes the state.

Reply:
Reset the position or use shallow copy.

Reply:
Actually, it shouldn't be a deep copy. Let me fix it.

Reply:
Fixed.